File: vb/Basic/Server.DataTypes/ManufacturingOrder.vb
'Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

Imports Opc.UaFx

Namespace DataTypes
    'Use the 'OpcDataTypeAttribute' to declare the type as a data type used in OPC UA as well.
    <OpcDataType("ns=2;s=ManufacturingOrder")>
    <OpcDataTypeEncoding("ns=2;s=ManufacturingOrder.Binary", Type:=OpcEncodingType.Binary)>
    Public Class ManufacturingOrder
        Private _jobs As MachineJob()

        Public Property Article As String

        Public Property Order As String

        Public Property JobsLength As Integer
            Get
                Return _jobs.Length
            End Get
            Set(ByVal value As Integer)
                Array.Resize(_jobs, value)
            End Set
        End Property

        'Use the 'OpcDataTypeMemberLengthAttribute' to define the length of a variable array
        'field using the value of another field.
        <OpcDataTypeMemberLength(NameOf(JobsLength))>
        Public Property Jobs As MachineJob()
            Get
                Return _jobs
            End Get
            Set(ByVal value As MachineJob())
                _jobs = value
            End Set
        End Property
    End Class
End Namespace

File: cs/Basic/Server.DataTypes/ManufacturingOrder.cs
// Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

namespace DataTypes
{
    using System;
    using Opc.UaFx;

    // Use the 'OpcDataTypeAttribute' to declare the type as a data type used in OPC UA as well.
    [OpcDataType("ns=2;s=ManufacturingOrder")]
    [OpcDataTypeEncoding("ns=2;s=ManufacturingOrder.Binary", Type = OpcEncodingType.Binary)]
    public class ManufacturingOrder
    {
        private MachineJob[] jobs;


        public string Article
        {
            get;
            set;
        }

        public string Order
        {
            get;
            set;
        }


        public int JobsLength
        {
            get => this.Jobs.Length;
            set => Array.Resize(ref this.jobs, value);
        }

        // Use the 'OpcDataTypeMemberLengthAttribute' to define the length of a variable array
        // field using the value of another field.
        [OpcDataTypeMemberLength(nameof(JobsLength))]
        public MachineJob[] Jobs
        {
            get => this.jobs;
            set => this.jobs = value;
        }
    }
}
